Add migration with boot pattern test

ldoktor opened the pull request:
This test starts the VM and starts migrations until a boot pattern is
reached. Optionally it allows sending another command (to resume
booting) and verifies the machine can ran properly.

One variant is added which makes pseries machines boot into slof, where
migration previously used to fail.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added a new migration test variant with boot pattern detection capabilities. The test enables comprehensive KVM migration validation during system boot phases with configurable timeouts and custom post-boot commands. Includes specialized support for POWER architectures with SLOF initialization control and autoboot parameter configuration.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/autotest/tp-qemu/pull/1082?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ldoktor committed:
Add migration with boot pattern test

This test starts the VM and starts migrations until a boot pattern is
reached. Optionally it allows sending another command (to resume
booting) and verifies the machine can ran properly.

One variant is added which makes pseries machines boot into slof, where
migration previously used to fail.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor commented:
@clebergnu would you please take a look at this?

clebergnu commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+    """
+    def wait_for_boot_pattern(boot_pattern, vm):
+        """
+        Waits till last_line_matches and survives vm-migration
```
This docstring line is "interesting": it uses a super non-formal version of "until", and then uses underscores for "last_line_matches".  Did you intend to use the aexpect function name (`read_until_last_line_matches()`)?  I suggest just using an explanation instead: "Waits until the last line matches the boot pattern, while surviving migration".

ldoktor replied:
Sure

clebergnu commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+                  logging.info)
+    end = time.time() + boot_pattern_timeout
+    vm.create()
+    bg_job = vm.serial_console.read_until_last_line_matches
```
I don't see the reason for this line.  A leftover?

ldoktor replied:
yep, I'll remove it...

luckyh replied:
+1

clebergnu reviewed (changes requested):
There's one line with unnecessary code.  Other than that, I have the following questions/considerations:

1) I see your point about having the `with_boot_pattern` and then the `in_slof` variant. I wonder if you could provide another variant, maybe with x86_64 and JeOS to really make this useful to more people.

2) I found the fact that the loop and thus the number of migrations not being deterministic is not a good thing, like the 3rd point in the test description ("Issue one or more migration after the pattern is reached").

3) I need to test this.  I got myself a Power machine (for a few hours) and I'll let you know how it goes.

ldoktor commented:
Hello @clebergnu one think I don't understand is the x86 variant. This is a power specific test focusing on existing BZ. That is why I created this test I don't know whether and what to focus on in x86 world but if someone needs such test they can re-use this test and configure it accordingly. Note that there is `migreate.with_reboot` which is being used regularly but that one does not make sure it migrates while booting firmware on ppc64 (so it did not trigger this BZ).

ldoktor commented:
As for the number of migration, it doesn't matter how many. It's as many as possible but at least one (before and after).

ldoktor commented:
And to the JeOS I also don't understand. This test does not filter-out JeOS, it should work perfectly there...

ldoktor commented:
@clebergnu are the questions addressed? Should I update this PR or do you still need some clarifications? I wanted to include this test in CI 1/2 a year ago...

luckyh commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+    boot_pattern_timeout = float(params.get("boot_pattern_timeout", 3600))
+    boot_pattern = params["boot_pattern"]
+    post_boot_pattern_cmd = (params.get("post_boot_pattern_cmd")
+                             .decode("unicode_escape"))
```
To avoid `AttributeError`, it should be either assigned with the default value `""` or used as a mandatory.

ldoktor replied:
yep

luckyh commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+    boot_pattern = params["boot_pattern"]
+    post_boot_pattern_cmd = (params.get("post_boot_pattern_cmd")
+                             .decode("unicode_escape"))
+    login_timeout = int(params.get("login_timeout", 360))
```
Not suggest to customize `login_timeout` if there is no special needs.

ldoktor replied:
Well this test is supposed to be used for boot-pattern checks. It might be useful to change the boot timeout __after__ the boot pattern is hit (it could be later in the boot process and we might want shorter timeout, or even longer one than default).

luckyh commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+        Waits till last_line_matches and survives vm-migration
+        """
+        patterns = [boot_pattern]
+        while True:
```
The thread would not be terminated when `boot_pattern_timeout` expired if really nothing could be got from `vm.serial_console`, then this test would be stuck until `test_timeout` expired.

ldoktor replied:
It should terminate after `boot_pattern_timeout` because of the `assert time.time() < end`. The reading (and matching) is done in a different thread so it should not interfere with this loop.

luckyh commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+            try:
+                if vm.serial_console.read_until_last_line_matches(patterns,
+                                                                  timeout=1):
+                    break
```
I think here should be `return True`, otherwise line 71 would always raise exception.

ldoktor replied:
Oh True, and the condition there is actually bad...

luckyh commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+    error.context("Start VM and wait for pattern while migrating",
+                  logging.info)
+    end = time.time() + boot_pattern_timeout
+    vm.create()
```
I'd suggest:
```patch
- start_vm = no
+ paused_after_start_vm = yes

- vm.create()
+ vm.resume()
```
How do you think about it?

ldoktor replied:
What would be the benefit? The `create` is used in many tests.

luckyh reviewed (changes requested):
There are several issues needed to be solved.

ldoktor commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,84 @@
+import time
+
+from autotest.client.shared import utils
+from autotest.client.shared import error
```
True, will fix..

ldoktor commented:
Fixed:
* rebased
* use avocado/virttest utils rather than autotest ones
* return True in the interrupted thread and related condition to detect bg thread failure
* provided default for "post_boot_pattern_cmd)

Not fixed:
* https://github.com/autotest/tp-qemu/pull/1082/commits/2a126a04240c1de12a3376056d62bf4a511cbf63#r210479813
* https://github.com/autotest/tp-qemu/pull/1082/commits/2a126a04240c1de12a3376056d62bf4a511cbf63#r210478291
* https://github.com/autotest/tp-qemu/pull/1082/commits/2a126a04240c1de12a3376056d62bf4a511cbf63#r210481449

luckyh commented on qemu/tests/migration_with_boot_pattern.py:
```diff
@@ -0,0 +1,87 @@
+                          logging.info)
+    end = time.time() + boot_pattern_timeout
+    vm.create()
+    bg_job = vm.serial_console.read_until_last_line_matches
```
Unreferenced var.

ldoktor replied:
oups, again a leftover from previous version...

ldoktor commented:
Changes:

* removed unused variable
* reordered imports to avoid pylint warning

Tested on x86 using tcg: `avocado --show all run migrate..in_slof..tcp..default --vt-arch ppc64 --vt-machine-type pseries --vt-qemu-bin /usr/local/bin/qemu-system-ppc64 --vt-qemu-dst-bin /usr/local/bin/qemu-system-ppc64  --vt-extra-params enable_kvm=no`

luckyh reviewed (dismissed):
LGTM

ldoktor commented:
@clebergnu would you please review this new version?

ldoktor commented:
@clebergnu, it's been a year since your last comment, can you please ack/nack or withdraw your review?

crobinso reviewed (commented):
LGTM
